[PG-1637] Remove any existing keys for tables on create

AndersAstrand opened the pull request:
Some crash scenarios can leave keys behind in the relation key file for OIDs that are unused. If these OIDs are later used when creating a non-encrypted relation the key has to be removed or we will do the wrong thing since the existence of a key makes us assume the relation is encrypted.

dAdAbird commented on contrib/pg_tde/src/pg_tde.c:
```diff
@@ -93,6 +93,7 @@ _PG_init(void)
 
 	check_percona_api_version();
 
+	pg_tde_init_data_dir();
```
what was the reason to move it here? I'm not objecting it, just curious

AndersAstrand replied:
As I mentioned in the commit message, this directory is used by the SMgr, and the SMgr is loaded and used regardless of whether any database have run CREATE EXTENSION. So the directory should be created regardless.

Practically it was because the call to pg_free_key_map_entry() crashed when the directory didn't exist :).

dAdAbird replied:
Yes, sorry, I missed it. Thanks

jeltz commented on contrib/pg_tde/src/smgr/pg_tde_smgr.c:
```diff
@@ -344,26 +344,47 @@ tde_mdcreate(RelFileLocator relold, SMgrRelation reln, ForkNumber forknum, bool
+		InternalKey *key = tde_smgr_get_key(&reln->smgr_rlocator);
 
-		if (key)
+		if (unlikely(key))
```
When can this even happen? Is it a thing which only happens if there are leftover keys or we are doing a recovery/replay on the replica? This branch should be explained.

AndersAstrand replied:
New version up that is much more strict about what can happen in each case.

jeltz reviewed (changes requested):
My first thought is: why can't we always delete keys, even for encrypted relations? And I do not find an answer in the PR.

AndersAstrand commented:
> My first thought is: why can't we always delete keys, even for encrypted relations? And I do not find an answer in the PR.

I have to think about this. I was worried about the recovery case, but there will be a WAL record for creating the key after this function in called, so it shouldn't be an issue. That would also remove the need for the branch you asked about I'm pretty sure. I need to verify though.

Commit:
Make tde_mdcreate more strict in it's behavior

Only create keys when MAIN fork is created, and trust
tde_smgr_should_encrypt() to know when to encrypt.

Also trust that the key has already been created if we're in recovery or
replication.

Commit:
Move WAL logging of relation key creation to smgr

Instead of it being done down in tdemap code controlled with a boolean,
we just do it where the decision to do it is made instead.

Commit:
Delete relation keys before create

There are crash scenarios where keys are left behind in the key file
even though the OID for the table goes unused. This meant that we could
have keys laying around for newly created plaintext relations after OID
wraparound.

Simply removing any existing keys when relations are created seems
appriopriate.

Also move creation of pg_tde data dir to library init. This directory is
used by the SMgr which is loaded regardless of whether any database are
yet to create extension or not.
